### PR TITLE
Adding Performance Tracking

### PR DESCRIPTION
**Pull request by khosravipasha**
Modified the command line client to be able to generate random user, conversation, and messages to simulate having many user.

One the server side, added a basic performance tracking and logging average time per operation.

Future work:
1. Adding a script that reads the performance logs, and generate graphs.
2. Make it easier to run several clients to be able to add more request on server
3. Expand on performance tracking to be able to find bottlenecks.

### CODE REVIEW

File: src/codeu/chat/client/ClientConversation.java [hunk 1]
```diff
@@ -68,6 +60,27 @@ static public boolean isValidTitle(String title) {
     return clean;
   }
 
+  // Print Conversation.  User context is used to map from owner UUID to name.
+  public static void printConversation(ConversationSummary c, ClientUser userContext) {
+    if (c == null) {
+      System.out.println("Null conversation");
+    } else {
+      final String name = (userContext == null) ? null : userContext.getName(c.owner);
```
Comment (evansenter) [inline src/codeu/chat/client/ClientConversation.java]: You should take a look at `Optional` in Java. http://www.oracle.com/technetwork/articles/java/java8-optional-2175753.html

File: src/codeu/chat/client/ClientConversation.java [hunk 2]
```diff
@@ -103,7 +122,24 @@ public void startConversation(String title, Uuid owner) {
...
   }
 
-  public void setCurrent(ConversationSummary conv) { currentSummary = conv; }
+  public Uuid startConversationAndGetUuid(String title, Uuid owner) {
+    final boolean validInputs = isValidTitle(title);
+
+    final Conversation conv = (validInputs) ? controller.newConversation(title, owner) : null;
+
+    if (conv == null) {
+      System.out.format("Error: conversation not created - %s.\n",
```
Comment (evansenter) [inline src/codeu/chat/client/ClientConversation.java]: LOG.error?

File: src/codeu/chat/client/ClientConversation.java [hunk 3]
```diff
@@ -103,7 +122,24 @@ public void startConversation(String title, Uuid owner) {
...
+          (validInputs) ? "server failure" : "bad input value");
+      return null;
+    } else {
+      LOG.info("New conversation: Title= \"%s\" UUID= %s", conv.title, conv.id);
+
+      currentSummary = conv.summary;
+
+      updateAllConversations(currentSummary != null);
```
Comment (evansenter) [inline src/codeu/chat/client/ClientConversation.java]: Why would conv.summary be null here ever?

File: src/codeu/chat/client/ClientUser.java [hunk 4]
```diff
@@ -57,6 +54,18 @@ static public boolean isValidName(String userName) {
     return clean;
   }
 
+  /* TODO remove later
```
Comment (evansenter) [inline src/codeu/chat/client/ClientUser.java]: Gut this.

File: src/codeu/chat/client/ClientUser.java [hunk 5]
```diff
@@ -57,6 +54,18 @@ static public boolean isValidName(String userName) {
...
+  public static String getUserInfoString(User user) {
+    return (user == null) ? "Null user" :
+        String.format(" User: %s\n   Id: %s\n   created: %s\n", user.name, user.id, user.creation);
+  }
+  */
+
+  // Move to User's toString()
```
Comment (evansenter) [inline src/codeu/chat/client/ClientUser.java]: Can this be done now?

File: src/codeu/chat/client/ClientUser.java [hunk 6]
```diff
@@ -102,6 +111,22 @@ public void addUser(String name, String password) {
     }
   }
 
+  public Uuid addUserAndGetUuid(String name) {
+    final boolean validInputs = isValidName(name);
+
+    final User user = (validInputs) ? controller.newUser(name) : null;
+
+    if (user == null) {
+      System.out.format("Error: user not created - %s.\n",
```
Comment (evansenter) [inline src/codeu/chat/client/ClientUser.java]: LOG.error?

File: src/codeu/chat/util/RandomGenerator.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package codeu.chat.util;
+
+// TODO used and modified some outside code, was not sure how to cite it, this was one suggested way
```
Comment (evansenter) [inline src/codeu/chat/util/RandomGenerator.java]: This is fine, you can remove this TODO.

**Comment by evansenter** [comment]
In general it looks good. As a high-level suggestion, you do *a lot* of juggling with `null`, checking if a value is assigned or not, and conditional logic contingent on the nullity of various variables. This tends to be risky code, as it's hard to follow what can, and can not be null. Java 8 adds support for `Optional` which I linked in the comments above, which leverages the type system to circumvent a large number of these `NullPointerException` (NPE)-like errors. The general idea is that you'd do something like:

```private ConversationSummary currentSummary = Optional.empty();

// ...

public static void printConversation(Optional<ConversationSummary> c, Optional<ClientUser> userContext) {```

...and then everywhere where you were before checking for `null` you would now use the interface methods, such as `isPresent`, `orElse`, etc. Check out the article, it's a good resource.

File: src/codeu/chat/util/RandomGenerator.java (hunk 7 above)
Comment (khosravipasha) [inline src/codeu/chat/util/RandomGenerator.java]: Done.

File: src/codeu/chat/client/ClientUser.java (hunk 6 above)
Comment (khosravipasha) [inline src/codeu/chat/client/ClientUser.java]: I tried to imitate other part of the code (the "addUser" method) and it was using System.out.format. It makes more sense to do the log, so I changed both to LOG.error.

File: src/codeu/chat/client/ClientConversation.java (hunk 2 above)
Comment (khosravipasha) [inline src/codeu/chat/client/ClientConversation.java]: I tried to imitate other part of the code and it was using System.out.format. It makes more sense to do the log, so I changed both to LOG.error.

File: src/codeu/chat/client/ClientConversation.java (hunk 1 above)
Comment (khosravipasha) [inline src/codeu/chat/client/ClientConversation.java]: Oh Java8 has many more new good feature, got to update my Java skills.
